### PR TITLE
Migrate project to BothamNetworking.

### LINKED ISSUE #5: Limit API param is not used as should
@flipper83 has noticed that the ``limit`` param used in the API is associated to the number of items you want to retrieve and not from the last item you want to retrieve. We should change this in the client side api to use ``size`` or ``count`` as named param instead of ``limit`` and also change a precondition we had related to this param.

### LINKED ISSUE #8: Integrate BothamNetworking

### LINKED ISSUE #9: Proposal: extension that add encoding of HTTPResponse body
We can have an extension for Result that give use directly SwiftyJSON format.

```Swift
bothamAPIClient.GET(anyPath) { result in
    result.mapJSON { json in
         return parseComics(json)
    }
}

// OR

bothamAPIClient.GET(anyPath) { result in
    result.mapJSON(parseComics)
}
```

I didn't test the code below(neither compiled), it's more like pseudo-code.
```Swift
import Result
import SwiftyJSON
extension ResultType where Value == HTTPResponse {

    public func mapJSON<U>(@noescape transform: JSON -> U) -> Result<U, Error> {
        return flatMap { dataToJSONResult($0)) }.map { transform($0) }
    }

    private func dataToJSONResult(data: NSData) -> Result<JSON, NSError> {
        Result.tryMap {
            let object: AnyObject = try NSJSONSerialization.JSONObjectWithData(data, options: .AllowFragments)
            return JSON(object)
        }
    }
}
```

### PR DESCRIPTION
**Pull request by pedrovgs**
Fix #9 fix #8 fix #5.

This PR is not going to compile or pass the tests in Travis environments because BothamNetworking has not been published yet. I've changed the project repository to a private on to be able to merge this branch into master and the rest of the pending branches before the first release even if the project is not passing in travis.

We should create a new target inside the BothamNetworking project with the SwiftyJson dependency and the json parsing implementation based on Result extension. What do you think?

### CODE REVIEW

File: MarvelApiClient/Result.swift [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+//  Created by Pedro Vicente Gomez on 05/01/16.
+//  Copyright © 2016 GoKarumi S.L. All rights reserved.
+//
+
+import Foundation
+import BothamNetworking
+import Result
+import SwiftyJSON
+
+extension ResultType where Value == HTTPResponse {
```
Comment (pedrovgs) [inline MarvelApiClient/Result.swift]: Ahhhhhh I was trying with Result instead of ResultType!!! Thanks Davide! Your Swift-Fu is over 9000!

Over9000

File: MarvelApiClient/CharactersAPIClient.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    init(apiClient: BothamAPIClient, parser: CharactersParser) {
+        self.parser = parser
+        self.apiClient = apiClient
+    }
+
+    public func getAll(offset offset: Int, limit: Int,
+        completion: (Result<GetCharactersDTO, BothamAPIClientError>) -> ()) {
+        assert(offset >= 0 && limit >= 0)
+        let params: [String:String?] =  [MarvelAPIParams.offset : "\(offset)", MarvelAPIParams.limit : "\(limit)"]
```
Comment (pedrovgs) [inline MarvelApiClient/CharactersAPIClient.swift]: I remember why we changed this. This is the ``Alamofire`` implementation:

```swift
// MARK: - Request

    /**
        Creates a request for the specified method, URL string, parameters, parameter encoding and headers.
        - parameter method:     The HTTP method.
        - parameter URLString:  The URL string.
        - parameter parameters: The parameters. `nil` by default.
        - parameter encoding:   The parameter encoding. `.URL` by default.
        - parameter headers:    The HTTP headers. `nil` by default.
        - returns: The created request.
    */
    public func request(
        method: Method,
        _ URLString: URLStringConvertible,
        parameters: [String: AnyObject]? = nil,
        encoding: ParameterEncoding = .URL,
        headers: [String: String]? = nil)
        -> Request
    {
        let mutableURLRequest = URLRequest(method, URLString, headers: headers)
        let encodedURLRequest = encoding.encode(mutableURLRequest, parameters: parameters).0
        return request(encodedURLRequest)
    }
```

Right now I'd like to change the API to use [String:String] instead of [String:String?] but this is just my opinion and the code is working. This is something we can discuss with the rest of the team.

File: MarvelAPIClientPlayground.playground/contents.xcplayground [hunk 3]
```diff
@@ -1,4 +1,4 @@
 <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
-<playground version='5.0' target-platform='ios' display-mode='rendered'>
```
Comment (pedrovgs) [inline MarvelAPIClientPlayground.playground/contents.xcplayground]: Sorry, I forgot to revert this change. Fixed in my last commit.

File: MarvelApiClient.xcworkspace/xcshareddata/MarvelApiClient.xcscmblueprint [hunk 4]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (pedrovgs) [inline MarvelApiClient.xcworkspace/xcshareddata/MarvelApiClient.xcscmblueprint]: Fixed!

File: MarvelApiClient/DefaultHeadersRequestInterceptor.swift [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+//  MarvelAPIClient
+//
+//  Created by Pedro Vicente Gomez on 04/01/16.
+//  Copyright © 2016 GoKarumi S.L. All rights reserved.
+//
+
+import Foundation
+import BothamNetworking
+
+class DefaultHeadersRequestInterceptor: BothamRequestInterceptor {
```
Comment (pedrovgs) [inline MarvelApiClient/DefaultHeadersRequestInterceptor.swift]: My interceptor implementation or the interceptor mechanism?

File: MarvelApiClient/MarvelAPIAuthentication.swift [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        let timestamp = timeProvider.currentTimeMillis()
+        let privateKey = MarvelAPIClient.privateKey
+        let publicKey = MarvelAPIClient.publicKey
+        let hash = MarvelHashGenerator.generateHash(Int(timestamp), privateKey: privateKey, publicKey: publicKey)
+        let authParams: [String:String?] = [ MarvelAPIParams.timestamp : "\(timestamp)",
+            MarvelAPIParams.apiKey : publicKey,
+            MarvelAPIParams.hash : hash]
+        return request.appendingParameters(authParams)
+    }
+
```
Comment (pedrovgs) [inline MarvelApiClient/MarvelAPIAuthentication.swift]: I don't think so, we don't need it right now. The ``BothamAPIClient`` is already returning an ``BothamAPIClientError.HTTPResponseError`` with the status code if the authentication is not working.

File: MarvelApiClient/MarvelAPIClientConfig.swift [hunk 7]
```diff
@@ -8,9 +8,9 @@
 
 import Foundation
 
-struct MarvelApiClientConfig {
+struct MarvelAPIClientConfig {
 
     static let host = "http://gateway.marvel.com/v1/public/"
-    static let defaultHeaders = ["Accept" : "application/json"]
+    static let defaultHeaders = ["Accept" : "application/json", "Content-Type" : "application/json"]
```
Comment (pedrovgs) [inline MarvelApiClient/MarvelAPIClientConfig.swift]: Nop.

File: MarvelApiClient/Result.swift [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    func mapJSON<U>(@noescape transform: JSON -> U) -> Result<U, Error> {
+        return flatMap { result in
+            let data = self.value?.body
+            return dataToJSONResult(data).map { transform($0) }
+        }
+    }
+
+    private func dataToJSONResult(data: NSData?) -> Result<JSON, Error> {
+        return Result<JSON, Error> {
+            let object: AnyObject = try NSJSONSerialization.JSONObjectWithData(data ?? NSData(),
```
Comment (pedrovgs) [inline MarvelApiClient/Result.swift]: Done! Review my last commit, my implementation is a little bit different, but it is also working :)

File: MarvelApiClient/CharactersAPIClient.swift (hunk 2 above)
Comment (davideme) [inline MarvelApiClient/CharactersAPIClient.swift]: I'm fine if we change in the library if we consider as default behaviour, and we open a ticket to implement as a config, that:
// By Default
["var1" : ""] generate var1=
// and set 
["var1" : ""] generate var1

File: MarvelApiClient/DefaultHeadersRequestInterceptor.swift (hunk 5 above)
Comment (davideme) [inline MarvelApiClient/DefaultHeadersRequestInterceptor.swift]: Interceptor mechanism, that to set Headers you have yo implement an interceptor,

Maybe if we see that a recuring feature we can define DefaultHeadersRequestInterceptor as part of Botham, and have as param or method of HTTPClient.

File: MarvelApiClient/CharactersAPIClient.swift (hunk 2 above)
Comment (pedrovgs) [inline MarvelApiClient/CharactersAPIClient.swift]: Can you open an issue describing this?

File: MarvelApiClient/MarvelAPIAuthentication.swift (hunk 6 above)
Comment (davideme) [inline MarvelApiClient/MarvelAPIAuthentication.swift]: Yeah, we but have to review that, we need a way to capture authentication error.

File: MarvelApiClient/DefaultHeadersRequestInterceptor.swift (hunk 5 above)
Comment (pedrovgs) [inline MarvelApiClient/DefaultHeadersRequestInterceptor.swift]: IMHO the interceptor mechanism is quite cool :) As you say, we can provide this implementation by default as part of BothamNetworking.

File: MarvelApiClient/MarvelAPIAuthentication.swift (hunk 6 above)
Comment (pedrovgs) [inline MarvelApiClient/MarvelAPIAuthentication.swift]: With the current BothamNetworking code we can implement all the "retry request to authenticate" mechanism from the BothamAPIClient users. If we want to implement this mechanism inside interceptors we will have to improve interceptors implementation to follow a chain of responsibility pattern. Fortunately we don't need it for this marvel api client :)

File: MarvelApiClient/MarvelAPIClientConfig.swift (hunk 7 above)
Comment (davideme) [inline MarvelApiClient/MarvelAPIClientConfig.swift]: Ok, we have something inconsistent.
https://github.com/Karumi/BothamNetworking/blob/34feaa97c2815c60d4b4d5035c9a9b860558f9a0/BothamNetworking/HTTPClient.swift#L32

We force check that the return content is JSON, but we don't send "Accept" : "application/json".

And for the Request Encoding+Content-Type maybe we should be able to configure it or provide one by default.

File: MarvelApiClient/MarvelAPIAuthentication.swift (hunk 6 above)
Comment (davideme) [inline MarvelApiClient/MarvelAPIAuthentication.swift]: let's skip this for now. we will review it when implementing authentication.

File: MarvelApiClient/MarvelAPIClientConfig.swift (hunk 7 above)
Comment (pedrovgs) [inline MarvelApiClient/MarvelAPIClientConfig.swift]: I've removed that check from the BothamNetworking project. This is the PR https://github.com/Karumi/BothamNetworking/pull/23

Comment (davideme) [inline MarvelApiClient/MarvelAPIClientConfig.swift]: As discussed outside of this PR we will remove the check from the BothamNetworking.

**Comment by pedrovgs** [comment]
@davideme as you suggested I've updated ``DefaultHeadersRequestInterceptor`` to do change any header value previously configured in the ``HTTPRequest`` creation. You can take a look at the new implementation in the test I've added heer ``DefaultHeadersRequestInterceptorTests``

**Comment by pedrovgs** [comment]
@davideme I've updated this PR to match with the latests BothamNetworking changes. If I'm not wrong, this is ready to merge.

**Comment by davideme** [comment]
:+1:
